### PR TITLE
Correct `scipy.optimize.linear_sum_assignment` usage

### LINKED ISSUE #4388: Scipy linear sum assignment is working correctly but lap.lapjv works not correctly.
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection, Other

### Bug

I find something wrong with the linear_sum_assignment usage in [https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/utils/matching.py#L61](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/utils/matching.py#L61)

the issuse opened in scipy [https://github.com/scipy/scipy/issues/19059](https://github.com/scipy/scipy/issues/19059)

what I find is the lap.lapjv  work properly, and the **scipy.optimize.linear_sum_assignment** works properly too. both are correct.  the scipy's version is **1.11.1**,  the lap's version is **0.5.4**. the wrong place is the codes of using the result of scipy.

- YOLO: [Ultralytics YOLOv8.0.21 🚀](https://github.com/ultralytics/ultralytics)      torch-2.0.1+cpu
- OS: Win10
- Python 3.10.12

### the matrix example

firstly clone the repo [https://github.com/ultralytics/ultralytics](https://github.com/ultralytics/ultralytics), and then using ultralytics to tracking.

```
import sys
sys.path.append(r'C:\Users\10696\Desktop\access\KalmanFilter\ultralytics')
from ultralytics import YOLO
model = YOLO('yolov8n.pt')
results = model.track(source=r"F:\MOT\MOT16-06-raw.webm", conf=0.3, iou=0.5, show=True)
```
[F:\MOT\MOT16-06-raw.webm](https://github.com/scipy/scipy/assets/34758215/abec2023-2546-460d-8eed-139ab1bc09b3)

I will debug into [ultralytics/trackers/utils/matching.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/utils/matching.py#L20)

this is the debug codes revised by myself.
```
def linear_assignment(cost_matrix, thresh, use_lap=False):
    """Linear assignment implementations with scipy and lap.lapjv."""
    if cost_matrix.size == 0:
        return np.empty((0, 2), dtype=int), tuple(range(cost_matrix.shape[0])), tuple(range(cost_matrix.shape[1]))
    if use_lap:
        # https://github.com/gatagat/lap
        _, x, y = lap.lapjv(cost_matrix, extend_cost=True, cost_limit=thresh)
        matches = [[ix, mx] for ix, mx in enumerate(x) if mx >= 0 and cost_matrix[ix, mx] <= thresh]
        zz = []
        for i in matches:
            zz.append(cost_matrix[i[0], i[1]])
        unmatched_a = np.where(x < 0)[0]
        unmatched_b = np.where(y < 0)[0]
    else:
        # https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linear_sum_assignment.html
        x, y = scipy.optimize.linear_sum_assignment(cost_matrix)  # row x, col y
        matches = np.asarray([[x[i], y[i]] for i in range(len(x)) if cost_matrix[x[i], y[i]] <= thresh])
        zz = []
        for i in matches:
            zz.append(cost_matrix[i[0], i[1]])
        if len(matches) == 0:
            unmatched_a = list(np.arange(cost_matrix.shape[0]))
            unmatched_b = list(np.arange(cost_matrix.shape[1]))
        else:
            unmatched_a = list(set(np.arange(cost_matrix.shape[0])) - set(matches[:, 0]))
            unmatched_b = list(set(np.arange(cost_matrix.shape[1])) - set(matches[:, 1]))
    return matches, unmatched_a, unmatched_b
```

the first cost_matrix is showed below, it has six rows and five columns.
it is very simple to find that the minimum iou cost is [0.30292, 0.27259, 0.29889, 0.5569, 0.34354] in different row and column.
So let's find the difference between lap.lapjv and scipy.optimize.linear_sum_assignment.

```
cost_matrix = array([[          1,     0.30292,           1,           1,           1],
                                 [          1,           1,     0.27259,           1,           1],
                                 [    0.29889,           1,           1,           1,           1],
                                 [          1,           1,           1,           1,      0.5569],
                                 [          1,           1,           1,     0.34354,           1],
                                 [          1,           1,           1,           1,           1]], dtype=float32)
```

firstly, setting use_lap = True, the **zz**, x, y result of lap.lapjv below, so the result is correct.
```
x = array([ 1,  2,  0,  4,  3, -1])
y = array([2, 0, 1, 4, 3])
zz = [0.30292338, 0.27259243, 0.29888666, 0.5569032, 0.34353834]
```

secondly, setting use_lap = False, the **zz**, x, y result of scipy.optimize.linear_sum_assignment below, so the result is correct.
```
x=array([1, 2, 0, 4, 3], dtype=int64)
y=array([0, 1, 2, 3, 4], dtype=int64)
zz = array([    0.30292,     0.27259,     0.29889,      0.5569,     0.34354], dtype=float32)
```
So the difference is the result returned by scipy or lap.
let's see lap's document [https://github.com/gatagat/lap](https://github.com/gatagat/lap)

`
cost, x, y = lap.lapjv(C)
The function lapjv(C) returns the assignment cost (cost) and two arrays, x, y. If cost matrix C has shape N x M, then x is a size-N array specifying to which column is row is assigned, and y is a size-M array specifying to which row each column is assigned. For example, an output of x = [1, 0] indicates that row 0 is assigned to column 1 and row 1 is assigned to column 0. Similarly, an output of x = [2, 1, 0] indicates that row 0 is assigned to column 2, row 1 is assigned to column 1, and row 2 is assigned to column 0.
`

it this way, both are correct, I want to make sure what happen to using the **lap.lapjv** and pull request.

### Environment

- YOLO: [Ultralytics YOLOv8.0.21 🚀](https://github.com/ultralytics/ultralytics)      torch-2.0.1+cpu
- OS: Win10
- Python 3.10.12

### Minimal Reproducible Example

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by ZouJiu1**
related issue for this pull request 
[https://github.com/scipy/scipy/issues/19059](https://github.com/scipy/scipy/issues/19059)
[https://github.com/ultralytics/ultralytics/issues/4388](https://github.com/ultralytics/ultralytics/issues/4388)

I correct the tracking usage of the result of scipy.optimize.linear_sum_assignment
Scipy linear sum assignment is working correctly, but the usage is not correctly, so I revise the codes.

### Summary
🐛🚀🔀

Fixed a bug and improved the performance of the `linear_assignment` function in `matching.py`, which is used for bipartite matching in the SORT tracker. The new code checks the cost matrix values, uses set operations and numpy functions, and swaps the output order.

> _`linear_assignment`_
> _Faster, better matching now_
> _Winter bug is gone_

### Walkthrough
* Fix and improve bipartite matching algorithm in `linear_assignment` function ([link](https://github.com/ultralytics/ultralytics/pull/4390/files?diff=unified&w=0#diff-4653a6d4709a8bfea076671e911fd6d18d751f047e45ee0b2777a9e1308ca922L37-R51))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancement of the matching algorithm in the Ultralytics tracking utility.

### 📊 Key Changes
- 🔄 Re-integration and revision of the `scipy.optimize.linear_sum_assignment` method for assigning matches.
- 🛠️ Fixes and improvements in the match filtering based on cost thresholds.
- 🧹 Code cleanup, removing outdated comments that discourage the use of the scipy method.
- ➕ Added checks and handling for edge cases with zero matches.

### 🎯 Purpose & Impact
- 💡 Aims to provide a more robust matching process in object tracking by revising the linear assignment algorithm.
- ⏱ Potentially improves the speed and accuracy of tracking objects across video frames.
- 🏗 Builds a stronger foundation for developers to implement tracking, leading to more reliable performance in real-world applications.
- 🔍 Ensures that unmatched cases are still handled correctly, which might improve overall tracking stability.

### CODE REVIEW

**Comment by ZouJiu1** [comment]
I have read the CLA Document and I sign the CLA

**Comment by ZouJiu1** [comment]
the tests all pass with the setting of True or False using the video [F:\MOT\MOT16-06-raw.webm](https://github.com/scipy/scipy/assets/34758215/abec2023-2546-460d-8eed-139ab1bc09b3).

File: ultralytics/trackers/utils/matching.py [hunk 1]
```diff
@@ -34,19 +34,21 @@ def linear_assignment(cost_matrix, thresh, use_lap=True):
         return np.empty((0, 2), dtype=int), tuple(range(cost_matrix.shape[0])), tuple(range(cost_matrix.shape[1]))
 
     if use_lap:
+        # https://github.com/gatagat/lap
         _, x, y = lap.lapjv(cost_matrix, extend_cost=True, cost_limit=thresh)
-        matches = [[ix, mx] for ix, mx in enumerate(x) if mx >= 0]
+        matches = [[ix, mx] for ix, mx in enumerate(x) if mx >= 0 and cost_matrix[ix, mx] <= thresh]
```
Comment (glenn-jocher) [inline ultralytics/trackers/utils/matching.py:39]: Hey I saw you updated the lap usage here. It was my understanding that the `cost_limit=thresh` argument in the line above already rejects matches below threshold. 

So does the new code actually change anything here or is it simply meant to align it closer to the scipy usage?

Comment (ZouJiu1) [reply ultralytics/trackers/utils/matching.py]: I restore the changed line " and cost_matrix[ix, mx] <= thresh" and commit it.

**Comment by glenn-jocher** [comment]
@ZouJiu1 PR merged, thank you for the fix!

Now it may be possible to remove `lap` package completely from Ultralytics. I'll do some testing and see if this is an option in the future. It's always great to have less dependencies :)

**Pull request merged**